docs_src/design/ucr/Global Common Configuration.md

```diff
@@ -0,0 +1,169 @@
+      KeepAlive = "10" # Seconds (must be 2 or greater)
+      Retained = "false"
+      AutoReconnect = "true"
+      ConnectTimeout = "5" # Seconds
+      SkipCertVerify = "false"
+      # Additional Default NATS Specific options that need to be here to enable evnironment variable overrides of them
+      Format = "nats"
+      RetryOnFailedConnect = "true"
+      QueueGroup = ""
+      Durable = ""
+      AutoProvision = "true"
+      Deliver = "new"
+      DefaultPubRetryAttempts = "2"
+      Subject = "edgex/#" # Required for NATS Jetstram only for stream autoprovsioning
+    ```
+In the above example only the **PublishTopicPrefix**, **SubscribeTopic**, **SubscribeEnabled**, **MessageQueue.Topics** and **ClientId** settings have unique values to that of the default EdgeX deployment values. 
+
+!!! note
+    In Levski release App Services don't have the **MessageQueue** section and Core Command's is **MessageQueue.Internal**. These inconstancies will be rectified in EdgeX 3.0 so all EdgeX services have the same **MessageQueue** section specified in the same manner. Also in EdgeX 3.0, the **PublishTopicPrefix** and **SubscribeTopic** settings will be replaced by entries in **MessageQueue.Topics**.
+
+There are other similar common sections not shown shown above. As can be seen from the two examples above there is much duplication of configuration settings across all the EdgeX services. This gives rise to the need to have all these common duplicate configuration settings in a single global source. 
+
+### Existing solutions
+There are no existing solutions for global configuration for EdgeX since the current configuration implementation is specific to EdgeX.
+
+### Requirements
+
+#### General
+
+- Services shall be able reference a global common configuration in a manner that is flexible for use with and without the Service Provider
+- Common environment overrides shall be applied to global common configuration settings only once before the services receive the setting values
+
+    - This avoids the need for duplicate environment overrides on each service in the docker compose files
+
+- Services must be able to override any of the global common configuration settings with local service specific TOML values
+
+    !!! example - "Example Core Data specific **Writable.Telemetry** and **Service** configuration settings in local TOML file"
+        ```toml
+          [Writable.Telemetry]
+            [Writable.Telemetry.Metrics] # All service's metric names must be present in this list.
+            EventsPersisted = false
+            ReadingsPersisted = false
+
+        ...
+        
+        [Service]
+        Port = 59880
+        StartupMsg = "This is the Core Data Microservice"
+        ```
+
+- Environment variable overrides shall be applied in the same manner as they are today. 
+    - After loaded from file (prior to pushing into the the Configuration Provider) 
+    - After loaded from the Configuration Provider

```

Comment:
Due to the environment variable cannot be changed after the container creation, I believe the config from Configuration Provider should take priority.
Or, the config cannot be changed at runtime or by restarting the service if the environment variable is defined.

Reply:
@cloudxxx8  It would be better to state the above in terms of requirements, e.g. changing an environment variable MUST (not?) (have some effect)

Reply:
> I believe the config from Configuration Provider should take priority.

I previous ADR we decided that Env Overrides have precedence over all others.

If we change this, it has the effect that **any** overrides added after the service has populated the Config Provider will always be ignored. Users would have to know to also make the change in the Config Provider to match the newly add overrides or delete service's config from Config Provider so new override values are pushed into the Config Provider.

Currently only changed values in the Config Provider that have overrides are ignored on the next start-up. Those made to the `Writable` section do take effect until the next time the service is restart.